src/main/kotlin/org/rust/lang/core/psi/util/RustPsiNavigationExtensions.kt

```diff
@@ -0,0 +1,159 @@
+
+    // Find edge direct children of parent within selection.
+    // element1 has to be first leaf of left child, and element2 - last leaf of right child
+    val realStartOffset = element1.textRange.startOffset
+    val realEndOffset = element2.textRange.endOffset
+
+    element1 = element1.getTopmostParentInside(parent)
+    if (realStartOffset != element1.textRange.startOffset) return emptyArray()
+
+    element2 = element2.getTopmostParentInside(parent)
+    if (realEndOffset != element2.textRange.endOffset) return emptyArray()
+
+    // Now collect non-whitespace children of parent between (inclusive) element1 and element2
+    val elements = collectElements(element1, element2.nextSibling) { it !is PsiWhiteSpace }
+
+    // Finally check if found elements meet requirements, and return result
+    elements.forEachIndexed { idx, element ->
+        if (!(element is RustStmtElement
+            || (idx == elements.size - 1 && element is RustExprElement)
+            || element is PsiComment
+            )) return emptyArray()
+    }
+
+    return elements
+}
+
+
+/**
+ * Finds two edge leaf PSI elements within given range.
+ */
+private fun PsiFile.getElementRange(startOffset: Int, endOffset: Int): Pair<PsiElement, PsiElement>? {
+    val element1 = findElementAtIgnoreWhitespaceBefore(startOffset) ?: return null
+    val element2 = findElementAtIgnoreWhitespaceAfter(endOffset - 1) ?: return null
+
+    // Elements have crossed (for instance when selection was inside single whitespace block)
+    if (element1.textRange.startOffset >= element2.textRange.endOffset) return null
+
+    return element1 to element2
+}
+
+/**
+ * Finds a leaf PSI element at the specified offset from the start of the text range of this node.
+ * If found element is whitespace, returns its next non-whitespace sibling.
+ */
+private fun PsiFile.findElementAtIgnoreWhitespaceBefore(offset: Int): PsiElement? {
+    val element = findElementAt(offset)
+    if (element is PsiWhiteSpace) {
+        return findElementAt(element.getTextRange().endOffset)
+    }
+    return element
+}
+
+/**
+ * Finds a leaf PSI element at the specified offset from the start of the text range of this node.
+ * If found element is whitespace, returns its previous non-whitespace sibling.
+ */
+private fun PsiFile.findElementAtIgnoreWhitespaceAfter(offset: Int): PsiElement? {
+    val element = findElementAt(offset)
+    if (element is PsiWhiteSpace) {
+        return findElementAt(element.getTextRange().startOffset - 1)
+    }
+    return element
+}
+
+/**
+ * Finds child of [parent] of which given element is descendant.
+ */
+private fun PsiElement.getTopmostParentInside(parent: PsiElement): PsiElement {
+    if (parent == this) return this
+
+    var element = this
+    while (parent != element.parent) {
+        element = element.parent
+    }
+    return element
+}
+
+private fun collectElements(start: PsiElement, stop: PsiElement?, pred: (PsiElement) -> Boolean): Array<out PsiElement> {
```

Comment:
similar, but different :smile: 

I tried to introduce some small abstraction over range of PSI elements, but they are not comparable, so we are not able to use Kotlin's ranges :(